In [53]:
import pandas as pd
import numpy as np
from sklearn.neural_network import MLPRegressor
from sklearn.preprocessing import OneHotEncoder
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error

In [3]:
years = [15, 16]
taxi_df = {}
for i in [15,16]:
    taxi_df[i] = pd.read_csv("download/" + str(i) \
                             + "taxi_with_weather_std.csv",\
                                    sep = ",")

In [4]:
y_df = {}
X_raw_df = {}
for i in years:
    y_df[i] =  pd.DataFrame(taxi_df[i]['Count'])
    X_raw_df[i] = taxi_df[i].drop(columns = ['Pickup Community Area', \
                                             'Count', 'Trip Start Timestamp'])

In [7]:
weekdays = ['0','1','2','3','4','5','6']
enc = OneHotEncoder(categories=[weekdays])
encoded_df = {}
for i in years:
    enc.fit(X_raw_df[i].weekday.values.reshape(-1, 1))
    encoded_df[i] = pd.DataFrame(enc.transform(X_raw_df[i].weekday.values.reshape(-1, 1))\
                                 .toarray().astype(int))\
                                .rename(columns = {0: 'Monday', 1:'Tuesday', 2:'Wedneseday',\
                                3:'Thursday', 4:'Friday', 5:'Saturday',\
                                6:'Sunday'})

In [8]:
X_encode_df = {}
for i in years:
    X_encode_df[i] = pd.concat([X_raw_df[i].drop(columns=['weekday']), encoded_df[i]], axis=1)

In [9]:
X_mat = {}
y_mat = {}
for i in years:
    X_mat[i] = X_encode_df[i].values
    y_mat[i] = y_df[i].values

In [32]:
mlp = MLPRegressor(hidden_layer_sizes=(100, ), activation='relu', solver='adam', 
                   alpha=0.0001, batch_size='auto', learning_rate='constant', 
                   learning_rate_init=0.001, power_t=0.5, max_iter=200, shuffle=True, 
                   random_state=None, tol=0.0001, verbose=False, warm_start=False, momentum=0.9, 
                   nesterovs_momentum=True, early_stopping=False, validation_fraction=0.1, 
                   beta_1=0.9, beta_2=0.999, epsilon=1e-08, n_iter_no_change=10)

In [14]:
my_mlp = mlp.fit(X_mat[16], y_mat[16])

/Users/weiyudu/anaconda3/lib/python3.7/site-packages/sklearn/neural_network/multilayer_perceptron.py:1316: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/weiyudu/anaconda3/lib/python3.7/site-packages/sklearn/neural_network/multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


In [33]:
my_mlp_15 = mlp.fit(X_mat[15], y_mat[15])

/Users/weiyudu/anaconda3/lib/python3.7/site-packages/sklearn/neural_network/multilayer_perceptron.py:1316: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/weiyudu/anaconda3/lib/python3.7/site-packages/sklearn/neural_network/multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


In [17]:
mlp_pred16 = my_mlp.predict(X_mat[16])
mean_squared_error(y_true = y_mat[16], y_pred = mlp_pred16)

933.1869161027486

In [34]:
# test
mlp_pred16 = my_mlp_15.predict(X_mat[16])
mean_squared_error(y_true = y_mat[16], y_pred = mlp_pred16)

2287.2475856451524

In [35]:
# train
mlp_pred16 = my_mlp_15.predict(X_mat[15])
mean_squared_error(y_true = y_mat[15], y_pred = mlp_pred16)

1700.2963848074755

In [19]:
mlp_pred15 = my_mlp.predict(X_mat[15])
mean_squared_error(y_true = y_mat[15], y_pred = mlp_pred15)

2858.014617520188

In [22]:
my_mlp2 = mlp2.fit(X_mat[16],y_mat[16])

/Users/weiyudu/anaconda3/lib/python3.7/site-packages/sklearn/neural_network/multilayer_perceptron.py:1316: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [21]:
mlp2 = MLPRegressor(hidden_layer_sizes=(20, ), activation='relu', solver='adam', 
                   alpha=0.0001, batch_size='auto', learning_rate='adaptive', 
                   learning_rate_init=0.001, power_t=0.5, max_iter=2000, shuffle=True, 
                   random_state=None, tol=0.0001, verbose=False, warm_start=False, momentum=0.9, 
                   nesterovs_momentum=True, early_stopping=True, validation_fraction=0.1, 
                   beta_1=0.9, beta_2=0.999, epsilon=1e-08, n_iter_no_change=10)

In [23]:
mlp_pred15_2 = my_mlp2.predict(X_mat[15])
mean_squared_error(y_true = y_mat[15], y_pred = mlp_pred15_2)

3841.4632362305265

In [29]:
mlp_pred15 = my_mlp2.predict(X_mat[16])
mean_squared_error(y_true = y_mat[16], y_pred = mlp_pred15)

1529.0299285796768

In [26]:
mlp3 = MLPRegressor(hidden_layer_sizes=(100, ), activation='relu', solver='adam', 
                   alpha=0.0001, batch_size='auto', learning_rate='adaptive', 
                   learning_rate_init=0.001, power_t=0.5, max_iter=1000, shuffle=True, 
                   random_state=None, tol=0.0001, verbose=False, warm_start=False, momentum=0.9, 
                   nesterovs_momentum=True, early_stopping=False, validation_fraction=0.1, 
                   beta_1=0.9, beta_2=0.999, epsilon=1e-08, n_iter_no_change=10)

In [28]:
my_mlp3 = mlp3.fit(X_mat[15],y_mat[15])

/Users/weiyudu/anaconda3/lib/python3.7/site-packages/sklearn/neural_network/multilayer_perceptron.py:1316: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/weiyudu/anaconda3/lib/python3.7/site-packages/sklearn/neural_network/multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


In [30]:
mlp_pred15 = my_mlp3.predict(X_mat[16])
mean_squared_error(y_true = y_mat[16], y_pred = mlp_pred15)

1932.5019416918426

In [63]:
mlp_pred15 = my_mlp3.predict(X_mat[16])
mean_squared_error(y_true = y_mat[16], y_pred = mlp_pred15)


1932.5019416918426

In [96]:
my_sum = 0
for i in range(676368):
    my_sum += abs((y_mat[16][i][0] - mlp_pred15[i]) /mlp_pred15[i])
my_sum = my_sum / 676368 * 100
my_sum

158.16570303962948

In [83]:
mlp_pred15[89]

-6.470264956355319

In [90]:
i=676367
abs((y_mat[16][i][0] - mlp_pred15[i]) /y_mat[16][i][0])

/Users/weiyudu/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: RuntimeWarning: divide by zero encountered in double_scalars
  


inf

In [95]:
mlp_pred15[676367]

-17.5929710031653

In [36]:
mlp4 = MLPRegressor(hidden_layer_sizes=(100, ), activation='relu', solver='sgd', 
                   alpha=0.0001, batch_size='auto', learning_rate='adaptive', 
                   learning_rate_init=0.005, power_t=0.5, max_iter=1000, shuffle=True, 
                   random_state=None, tol=0.0001, verbose=False, warm_start=False, momentum=0.9, 
                   nesterovs_momentum=True, early_stopping=False, validation_fraction=0.1, 
                   beta_1=0.9, beta_2=0.999, epsilon=1e-08, n_iter_no_change=10)

In [38]:
my_mlp_4 = mlp4.fit(X_mat[15], y_mat[15])

/Users/weiyudu/anaconda3/lib/python3.7/site-packages/sklearn/neural_network/multilayer_perceptron.py:1316: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Users/weiyudu/anaconda3/lib/python3.7/site-packages/sklearn/neural_network/_base.py:194: RuntimeWarning: overflow encountered in square
  return ((y_true - y_pred) ** 2).mean() / 2
/Users/weiyudu/anaconda3/lib/python3.7/site-packages/sklearn/neural_network/multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


In [50]:
mlp_pred15 = my_mlp_4.predict(X_mat[16])
mean_squared_error(y_true = y_mat[16], y_pred = mlp_pred15)

ValueError: Input contains NaN, infinity or a value too large for dtype('float64').

In [49]:
np.nan_to_num(X_mat[16])

array([[ 0.562, -0.984,  0.179, ...,  0.   ,  0.   ,  0.   ],
       [ 0.562, -0.984,  0.179, ...,  0.   ,  0.   ,  0.   ],
       [ 0.562, -0.984,  0.179, ...,  0.   ,  0.   ,  0.   ],
       ...,
       [-0.704,  0.8  ,  0.6  , ...,  0.   ,  0.   ,  0.   ],
       [-0.704,  0.8  ,  0.6  , ...,  0.   ,  0.   ,  0.   ],
       [-0.704,  0.8  ,  0.6  , ...,  0.   ,  0.   ,  0.   ]])

## Training data - 2016, Testing data - 2015<br>
Model Default<br>
Training MSE - 933, Testing MSE - 2858<br>

Model 2<br>
Training MSE - 1529, Testing MSE - 3841<br>


## Training data - 2015, Testing data - 2016<br>
Model Default<br>
Training MSE - 1700, Testing MSE - 2287<br>

Model 3<br>
Training MSE - 1210, Testing MSE - 1932<br>